File: apz_core/plugins/contact_picker/v1.0.0/ios/Classes/ApzContactPicker.swift
import Flutter
import UIKit
import ContactsUI // Required for CNContactPickerViewController
import Contacts // Required for CNContact, CNContactFormatter, CNContactStore

public class ApzContactPicker: NSObject, FlutterPlugin, CNContactPickerDelegate {

  private var channel: FlutterMethodChannel!
  private var pendingResult: FlutterResult? // Holds the Flutter Result callback

  // MARK: - FlutterPlugin Registration

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "contact_picker_plugin", binaryMessenger: registrar.messenger())
    let instance = ApzContactPicker()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
    // No need for ActivityAware equivalent on iOS like Android.
    // CNContactPickerDelegate handles the UI presentation and dismissal.
  }

  // MARK: - MethodCallHandler (Flutter to Native)

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "pickContact" {
      self.pendingResult = result // Store the result callback for later use
      launchContactPicker()
    } else {
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Contact Picker Launch Logic

  private func launchContactPicker() {
    let contactPicker = CNContactPickerViewController()
    contactPicker.delegate = self // Set the plugin as the delegate to receive callbacks

    // Present the contact picker from the currently active UIViewController.
    // This is typically the rootViewController of the key window.
    if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
      rootViewController.present(contactPicker, animated: true, completion: nil)
    } else {
      pendingResult?(FlutterError(code: "VIEW_CONTROLLER_NOT_FOUND",
                              message: "Could not find a UIViewController to present the contact picker.",
                              details: nil))
      pendingResult = nil
    }
  }

  // MARK: - CNContactPickerDelegate Methods (Contact Picker Callbacks)

  // This method is called when the user successfully selects a contact.
  public func contactPicker(_ picker: CNContactPickerViewController, didSelect contact: CNContact) {
    picker.dismiss(animated: true, completion: nil) // Dismiss the picker UI

    // Extract details and send them back to Flutter.
    let contactMap: [String: Any?] = extractContactDetails(contact: contact)
    pendingResult?(contactMap)
    pendingResult = nil // Clear the pending result
  }

  // This method is called when the user cancels the contact picker.
  public func contactPickerDidCancel(_ picker: CNContactPickerViewController) {
    picker.dismiss(animated: true, completion: nil) // Dismiss the picker UI
    pendingResult?(nil) // Send nil back to Flutter to indicate cancellation
    pendingResult = nil // Clear the pending result
  }

  // MARK: - Helper to Extract Contact Details

  private func extractContactDetails(contact: CNContact) -> [String: Any?] {
    var contactInfo: [String: Any?] = [:]

    // Full Name
    let formatter = CNContactFormatter()
    formatter.style = .fullName
    if let fullName = formatter.string(from: contact) {
        contactInfo["fullName"] = fullName
    } else {
        contactInfo["fullName"] = nil
    }

    // Phone Number (get the first one if available)
    if let phoneNumber = contact.phoneNumbers.first?.value.stringValue {
      contactInfo["phoneNumber"] = phoneNumber
    } else {
      contactInfo["phoneNumber"] = nil
    }

    // Email (get the first one if available)
    if let email = contact.emailAddresses.first?.value as? String {
      contactInfo["email"] = email
    } else {
      contactInfo["email"] = nil
    }

    // Thumbnail Image Data
    // CNContact.thumbnailImageData is already an Optional<Data> (Data? in Swift).
    // If there's no thumbnail, it will be nil, which is directly handled.
    if let imageData = contact.thumbnailImageData {
        // Convert Swift's Data to Flutter's expected FlutterStandardTypedData (Uint8List on Dart side)
        contactInfo["thumbnail"] = FlutterStandardTypedData(bytes: imageData)
    } else {
        contactInfo["thumbnail"] = nil // Explicitly set to nil if no thumbnail
    }

    return contactInfo
  }
}

File: apz_core/plugins/contact_picker/v1.0.0/android/src/main/kotlin/com/iexceed/apz_contact_picker/ApzContactPicker.kt
package com.iexceed.apz_contact_picker

import android.app.Activity
import android.content.Intent
import android.database.Cursor
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.provider.ContactsContract
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

import java.io.ByteArrayOutputStream

/** ApzContactPicker */
class ApzContactPicker: FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener {

  private lateinit var channel : MethodChannel
  private var currentActivity: Activity? = null
  private var pendingResult: Result? = null
  private val PICK_CONTACT_REQUEST = 1

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "contact_picker_plugin")
    channel.setMethodCallHandler(this)
  }

  /// This method is called when Dart invokes a method on the MethodChannel.
  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "pickContact") {
      this.pendingResult = result
      // Directly launch the contact picker.
      // Permission check is now assumed to be handled on the Dart side before this call.
      launchContactPicker()
    } else {
      result.notImplemented()
    }
  }

  /// This method launches the Android system's contact picker UI.
  private fun launchContactPicker() {
    if (currentActivity == null) {
      pendingResult?.error("ACTIVITY_NOT_ATTACHED", "Plugin not attached to an activity.", null)
      pendingResult = null
      return
    }

    val intent = Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI)
    try {
      currentActivity?.startActivityForResult(intent, PICK_CONTACT_REQUEST)
    } catch (e: Exception) {
      pendingResult?.error("ACTIVITY_NOT_FOUND", "Could not launch contact picker: ${e.message}", null)
      pendingResult = null
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  // --- ActivityAware methods ---
  // These methods manage the plugin's lifecycle in relation to the Android Activity.
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    currentActivity = binding.activity
    binding.addActivityResultListener(this)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    currentActivity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    currentActivity = binding.activity
    binding.addActivityResultListener(this)
  }

  override fun onDetachedFromActivity() {
    currentActivity = null
  }

  // --- ActivityResultListener (for contact picker result) ---
  /// This method receives the result from an activity launched using startActivityForResult (e.g., the native contact picker).
  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
    if (requestCode == PICK_CONTACT_REQUEST) {
      if (pendingResult == null) {
        return false
      }

      if (resultCode == Activity.RESULT_OK) { // User successfully selected a contact
        val contactUri: Uri? = data?.data
        if (contactUri != null) {
          queryContactDetails(contactUri) // Proceed to extract details from the selected contact URI
        } else {
          pendingResult?.success(null) // User picked nothing/canceled somehow
          pendingResult = null
        }
      } else if (resultCode == Activity.RESULT_CANCELED) { // User canceled the picker
        pendingResult?.success(null) // Send null back to Dart to indicate cancellation.
        pendingResult = null
      } else { // Some other error occurred during picking.
        pendingResult?.error("PICK_FAILED", "Contact picker failed with result code: $resultCode", null)
        pendingResult = null
      }
      return true
    }
    return false
  }

  /// This method queries the contact details (name, phone, email, thumbnail) from the given contact URI and sends them back to Dart as a map.
  private fun queryContactDetails(contactUri: Uri) {
    val result: MutableMap<String, Any?> = mutableMapOf()
    var cursor: Cursor? = null
    try {
      // Get contact ID
      cursor = currentActivity?.contentResolver?.query(contactUri, arrayOf(ContactsContract.Contacts._ID), null, null, null)
      cursor?.moveToFirst()
      val contactId = cursor?.getString(cursor.getColumnIndexOrThrow(ContactsContract.Contacts._ID))

      if (contactId != null) {
        // Full Name
        val displayNameCursor = currentActivity?.contentResolver?.query(
          ContactsContract.Data.CONTENT_URI,
          arrayOf(ContactsContract.Data.DISPLAY_NAME),
          "${ContactsContract.Data.CONTACT_ID} = ? AND ${ContactsContract.Data.MIMETYPE} = ?",
          arrayOf(contactId, ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE),
          null
        )
        displayNameCursor?.use {
          if (it.moveToFirst()) {
            result["fullName"] = it.getString(it.getColumnIndexOrThrow(ContactsContract.Data.DISPLAY_NAME))
          }
        }

        // Phone Number
        val phoneCursor = currentActivity?.contentResolver?.query(
          ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
          arrayOf(ContactsContract.CommonDataKinds.Phone.NUMBER, ContactsContract.CommonDataKinds.Phone.TYPE),
          "${ContactsContract.CommonDataKinds.Phone.CONTACT_ID} = ?",
          arrayOf(contactId),
          null
        )
        phoneCursor?.use {
          if (it.moveToFirst()) { // Get the first phone number as primary
            result["phoneNumber"] = it.getString(it.getColumnIndexOrThrow(ContactsContract.CommonDataKinds.Phone.NUMBER))
          }
        }

        // Email (Optional)
        val emailCursor = currentActivity?.contentResolver?.query(
          ContactsContract.CommonDataKinds.Email.CONTENT_URI,
          arrayOf(ContactsContract.CommonDataKinds.Email.ADDRESS),
          "${ContactsContract.CommonDataKinds.Email.CONTACT_ID} = ?",
          arrayOf(contactId),
          null
        )
        emailCursor?.use {
          if (it.moveToFirst()) { // Get the first email as primary
            result["email"] = it.getString(it.getColumnIndexOrThrow(ContactsContract.CommonDataKinds.Email.ADDRESS))
          }
        }

        // --- FIX START ---
        // Thumbnail Image - Robust handling for missing images
        var thumbnailBytes: ByteArray? = null
        try {
            currentActivity?.contentResolver?.openAssetFileDescriptor(Uri.withAppendedPath(contactUri, ContactsContract.Contacts.Photo.CONTENT_DIRECTORY), "r")?.use { photoData ->
                val bitmap = BitmapFactory.decodeFileDescriptor(photoData.fileDescriptor)
                if (bitmap != null) {
                    val stream = ByteArrayOutputStream()
                    // Using JPEG for smaller size, 80% quality. Change to PNG if transparency is required.
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 80, stream)
                    thumbnailBytes = stream.toByteArray()
                    stream.close()
                }
            }
        } catch (e: Exception) {
            // Log the error but continue execution without a thumbnail.
            // Replace with Android's Log.e for production use:
            // android.util.Log.e("ApzContactPicker", "Error loading contact thumbnail: ${e.message}", e)
            println("ApzContactPicker: Error loading contact thumbnail: ${e.message}")
            thumbnailBytes = null // Ensure it's null on error
        }
        result["thumbnail"] = thumbnailBytes
        // --- FIX END ---
      }

      pendingResult?.success(result)
    } catch (e: Exception) {
      pendingResult?.error("CONTACT_DETAILS_ERROR", "Error retrieving contact details: ${e.message}", null)
    } finally {
      cursor?.close()
      pendingResult = null // Clear pending result after completion
    }
  }
}